### PR TITLE
Analyzer subquery support FINAL

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Analyzer support FINAL for subqueries. Example: `SELECT id, value FROM (SELECT id, value FROM test_table WHERE id = 0) FINAL`.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit f504e8d66ab00af5152791289e20357cd3b4b558 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64133/f504e8d66ab00af5152791289e20357cd3b4b558/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

**Review by alexey-milovidov — APPROVED** [review]
LGTM

**Comment by alexey-milovidov** [comment]
It should be applied only for tables that support FINAL.

File: src/Analyzer/Utils.cpp [hunk 1]
```diff
@@ -359,7 +359,28 @@ void addTableExpressionOrJoinIntoTablesInSelectQuery(ASTPtr & tables_in_select_q
...
+        else if (auto * table_node = table_or_table_function_node->as<TableNode>())
+            table_node->setOrMergeTableExpressionModifiers(table_expression_modifiers);
+        else if (auto * table_function_node = table_or_table_function_node->as<TableFunctionNode>())
+            table_function_node->setOrMergeTableExpressionModifiers(table_expression_modifiers);
+        else
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Expected table, table function or identifier node");
+    }
+}
+
+QueryTreeNodes extractAllTableReferences(const QueryTreeNodePtr & tree, bool extract_table_function_nodes, bool extract_identifier_nodes)
```
Comment (novikd) [inline src/Analyzer/Utils.cpp:383]: I think `extractAllTableReferences` should be used only for resolved query trees. It's better to write a separate function to traverse unresolved trees.

File: src/Analyzer/Resolve/QueryAnalyzer.cpp [hunk 2]
```diff
@@ -1122,24 +1122,38 @@ void QueryAnalyzer::validateTableExpressionModifiers(const QueryTreeNodePtr & ta
...
+
+    auto & table_expression_modifiers = table_node ? table_node->getTableExpressionModifiers() : table_function_node->getTableExpressionModifiers();
+    if (!table_expression_modifiers.has_value())
+        return;
+
+    const auto & storage = table_node ? table_node->getStorage() : table_function_node->getStorage();
+    if (table_expression_modifiers->hasFinal() && !storage->supportsFinal())
     {
-        auto table_expression_modifiers = table_node ? table_node->getTableExpressionModifiers() : table_function_node->getTableExpressionModifiers();
+        if (!table_expression_modifiers->isFromParentSubquery())
```
Comment (novikd) [inline src/Analyzer/Resolve/QueryAnalyzer.cpp:1135]: It will work incorrectly in the case:
```sql
SELECT
FROM (SELECT * FROM x FINAL) SAMPLE 0.1
```

**Review by novikd — COMMENTED** [review]
I suppose it's better to implement this feature on **resolved** query tree. I can store table modifiers in the current scope and apply them when you resolve Join Tree. Such an approach would make more sense.

**Comment by alexey-milovidov** [comment]
@kitaisreal, please address the review comments and add tests for the cases highlighted by @novikd

**Comment by alexey-milovidov** [comment]
Please open another PR when you continue working on this feature.

**Pull request closed without merging**